processing/src/main/java/org/apache/druid/query/DefaultQueryMetrics.java

```diff
@@ -24,50 +24,89 @@
 import org.apache.druid.query.filter.Filter;
 import org.joda.time.Interval;
 
+import javax.annotation.concurrent.GuardedBy;
+import java.util.Arrays;
 import java.util.HashMap;
 import java.util.List;
 import java.util.Map;
 import java.util.concurrent.TimeUnit;
 
 /**
- * DefaultQueryMetrics is unsafe for use from multiple threads. It fails with RuntimeException on access not from the
- * thread where it was constructed. To "transfer" DefaultQueryMetrics from one thread to another {@link #ownerThread}
- * field should be updated.
+ * A basic thread-safe implementation of the {@link QueryMetrics} interface.
+ *
+ * <p><b>Note for inheritence:</b> Subclass should override {@link #makeCopy()} method to return
+ * an instance of that subclass. See {@link org.apache.druid.query.groupby.DefaultGroupByQueryMetrics}
+ * for an example.
  */
 public class DefaultQueryMetrics<QueryType extends Query<?>> implements QueryMetrics<QueryType>
 {
-  protected final ObjectMapper jsonMapper;
-  protected final ServiceMetricEvent.Builder builder = new ServiceMetricEvent.Builder();
-  protected final Map<String, Number> metrics = new HashMap<>();
+  private static final EmittingLogger log = new EmittingLogger(DefaultQueryMetrics.class);
 
-  /** Non final to give subclasses ability to reassign it. */
-  protected Thread ownerThread = Thread.currentThread();
+  protected final ObjectMapper jsonMapper;
+  protected final Object lock = new Object();
+  @GuardedBy("lock") private final Map<String, String> singleValueDims = new HashMap<>();
+  @GuardedBy("lock") private final Map<String, String[]> multiValueDims = new HashMap<>();
+  @GuardedBy("lock") private final Map<String, Number> metrics = new HashMap<>();
 
   public DefaultQueryMetrics(ObjectMapper jsonMapper)
   {
     this.jsonMapper = jsonMapper;
   }
 
-  protected void checkModifiedFromOwnerThread()
+  // copy constructor, used by makeCopy()
+  public DefaultQueryMetrics(DefaultQueryMetrics that)
   {
-    if (Thread.currentThread() != ownerThread) {
-      throw new IllegalStateException(
-          "DefaultQueryMetrics must not be modified from multiple threads. If it is needed to gather dimension or "
-          + "metric information from multiple threads or from an async thread, this information should explicitly be "
-          + "passed between threads (e. g. using Futures), or this DefaultQueryMetrics's ownerThread should be "
-          + "reassigned explicitly");
-    }
+    this.jsonMapper = that.jsonMapper;
+    this.singleValueDims.putAll(that.singleValueDims);
+    this.multiValueDims.putAll(that.multiValueDims);
+    this.metrics.putAll(that.metrics);
   }
 
   protected void setDimension(String dimension, String value)
   {
-    checkModifiedFromOwnerThread();
-    builder.setDimension(dimension, value);
+    synchronized (lock) {
+      String oldValue = singleValueDims.put(dimension, value);
+      if (oldValue != null && !oldValue.equals(value)) {
```

Comment:
Equal values should not be tolerated. Any reassignment of already existing dimension or metric should be illegal.

Reply:
ServerManager.buildAndDecorateQueryRunner reassigns segment id in each MetricsEmittingQueryRunner. Do you think we should fix that or just tolerate it? I think allowing for equal values make no harm and is convenient in such cases.

Reply:
Allowing equal metrics could easily shallow mistakes. Metrics or dimensions could be equal because they happen to collide (especially in tests/staging environment), but logically they are different, that will hit in production.

Reply:
I agree that allowing equal metrics could shallow mistakes like multiple threads trying to report the same metrics, will prohibit such usage. But I think we should allow equal dimension values because there are valid use cases such as nesting MetricsEmittingQueryRunner. Do you think so?

Reply:
Could you give a specific example where setting the same dimensions from multiple threads is needed? Both #4826 and #5128 are due to concurrent metrics reporting, not dimensions setting.

I don't want to allow setting dimensions from multiple threads until it proves to be needed.

Reply:
It's possible to set dimensions from multiple threads. For example, MetricsEmittingQueryRunner adds custom dimension in one thread [#L91](https://github.com/apache/incubator-druid/blob/3ae563263a23000560749071d262727d47296856/processing/src/main/java/org/apache/druid/query/MetricsEmittingQueryRunner.java#L91) and other dimension in possibility another thread [#L112](https://github.com/apache/incubator-druid/blob/3ae563263a23000560749071d262727d47296856/processing/src/main/java/org/apache/druid/query/MetricsEmittingQueryRunner.java#L112).

It's also possible for one thread to set the same dimensions more than once. For instance, the [query runner](https://github.com/apache/incubator-druid/blob/359576a80bccaea4168e31323e931a806b26159b/server/src/main/java/org/apache/druid/server/coordination/ServerManager.java#L295) for historical  segment processing contains two MetricsEmittingQueryRunner, which would set "segment" and "status" dimension twice.

Reply:
I don't think there are cases for setting the same dimensions from multiple threads. But the overhead to prohibit such usages is a bit high for me. You needs to maintain the owner thread for each dimension. And I think the checks in reportMetric and emit are sufficient to catch most bugs.